[OpenXLA] Migrate to pull XLA from OpenXLA

ManfeiBai opened the pull request:
**[Description]**
PyTorch/XLA migrate to pull XLA from OpenXLA by replacing TensorFlow with OpenXLA after deprecating XRT usage, and replace TensorFlow-pin with OpenXLA-pin to May09

**[What has changed]**
1. change from `@org_tensorflow//tensorflow/compiler/xla/...` to `@xla//xla/...`
2. change from `@org_tensorflow//tensorflow/tsl/...` to `@tsl//tsl/...`
3. change from `#include "tensorflow/compiler/xla/...` to `#include "xla/...`
4. change from `#include "tensorflow/tsl/...` to `#include "tsl/...`
5. change from pull XLA from [TensorFlow](https://github.com/tensorflow/tensorflow) with `tf_patches` to pull XLA from [OpenXLA](https://github.com/openxla/xla) with `openxla_patches`

**[Start to do OpenXLA-pin]**
After this PR, we will update OpenXLA-pin instead of TensorFlow-pin. We used to update `WORKSPACE` and `tf_patches` to update used TensorFlow's version, aka TensorFlow-pin update. After this PR, we will update OpenXLA-pin by updating `WORKSPACE` and `openxla_patches` to update used OpenXLA's version.


Thanks, @miladm @JackCaoG @will-cromar @puneith @ddunl @stgpetrovic @wonjoolee95 @lsy323 @aspirido
According to https://github.com/pytorch/xla/pull/5202#discussion_r1237741669, we should give AWS folks a heads up when we merge this pr.

ManfeiBai commented:
`xla/service/gpu/runtime/executable.cc` skip one change

ManfeiBai commented:
skip `pprof.BUILD`

ManfeiBai commented:
found same error like https://github.com/pytorch/xla/pull/5202/commits/567657f4ccd7761ad1aa26ed585964060a76f1a0 in https://github.com/tensorflow/tensorflow/issues/58851#issuecomment-1384305876

miladm reviewed (commented):
Thanks. I took a quick pass at the PR. Will wait for the CI tests to pass and review.

@ManfeiBai please add a few lines in the PR description explaining the process of updating TorchXLA dependency to openXLA via pin-updates.

JackCaoG commented:
Hmm TPU CI seems to failed with some linker error
```
Step #2 - "build_pytorch-xla-test_docker_image":     ERROR: /src/pytorch/xla/BUILD:1:10: Linking _XLAC.so failed: (Exit 1): gcc failed: error executing command /usr/bin/gcc @bazel-out/k8-opt/bin/_XLAC.so-2.params
Step #2 - "build_pytorch-xla-test_docker_image":     /usr/bin/ld.gold: error: bazel-out/k8-opt/bin/_solib_k8/_U@local_Uconfig_Ucuda_S_Scuda_Ccudart___Ucuda_Scuda_Slib/libcudart.so: file is empty
Step #2 - "build_pytorch-xla-test_docker_image":     collect2: error: ld returned 1 exit status
```

ManfeiBai commented:
> Hmm TPU CI seems to failed with some linker error
> 
> ```
> Step #2 - "build_pytorch-xla-test_docker_image":     ERROR: /src/pytorch/xla/BUILD:1:10: Linking _XLAC.so failed: (Exit 1): gcc failed: error executing command /usr/bin/gcc @bazel-out/k8-opt/bin/_XLAC.so-2.params
> Step #2 - "build_pytorch-xla-test_docker_image":     /usr/bin/ld.gold: error: bazel-out/k8-opt/bin/_solib_k8/_U@local_Uconfig_Ucuda_S_Scuda_Ccudart___Ucuda_Scuda_Slib/libcudart.so: file is empty
> Step #2 - "build_pytorch-xla-test_docker_image":     collect2: error: ld returned 1 exit status
> ```

thanks, Jack, yeah, this error appear after adding `"@xla//xla/stream_executor:cuda_platform",` to BUILD's dep, is debuging this error now in https://github.com/pytorch/xla/pull/5209

JackCaoG commented:
hmm, for a TPUVM environment, we shouldn't need any cuda bits, maybe only include this dependency on cuda build?

ManfeiBai commented:
> hmm, for a TPUVM environment, we shouldn't need any cuda bits, maybe only include this dependency on cuda build?

Thanks, Jack, tested in https://github.com/pytorch/xla/pull/5209 with only include `"@xla//xla/stream_executor:cuda_platform",` in cuda build:
```
    ] + if_cuda_is_configured([
        "@xla//xla/stream_executor:cuda_platform",
    ]),
```

JackCaoG commented:
Thanks! ..I am curious why XRT CI also passed, I kind of expected it to fail

ManfeiBai commented:
> Thanks! ..I am curious why XRT CI also passed, I kind of expected it to fail

iiuc, XRT CI = non-XRT CI + XRT tests,

after we deleted almost all code of XRT, we deleted code added by `disable_xrt`, so we skipped all `disable_xrt` part;

now XRT CI is doing the same thing like non-XRT CI, so it passed too;

because OpenXLA don't support XRT, so there is no XRT code, so every build and tests build and passed without XRT, using OpenXLA

please correct me if I'm wrong, @will-cromar

JackCaoG commented:
Ok that make sense, this CI is doing more than I thought. My impression is we only disable the XRT but you actually deleted it in this pr. I guess it make sense since you want to delete some TPU topo stuff. I will try to review it tmr, but we can't merge this pr until we decide the future of the XRT. Great work, thanks @ManfeiBai !

JackCaoG commented on WORKSPACE:
```diff
@@ -23,71 +23,75 @@ python_configure(
-        "//tf_patches:triton_filesystem.diff",
-        "//tf_patches:xla_bzl.diff",
-        "//tf_patches:xplane.diff",
+        "//openxla_patches:absl_statusor.diff",
+        "//openxla_patches:cache_urls.diff",
+        "//openxla_patches:cuda_graph.diff",
+        "//openxla_patches:f16_abi_clang.diff",
+        "//openxla_patches:gpu_race_condition.diff",
+        "//openxla_patches:grpc_version.diff",
+        "//openxla_patches:optimized_function_graph.diff",
+        "//openxla_patches:profiler_trace.diff",
+        "//openxla_patches:stream_executor.diff",
+        "//openxla_patches:thread_local_random.diff",
+        "//openxla_patches:topk_rewriter.diff",
+        "//openxla_patches:triton_filesystem.diff",
+        "//openxla_patches:xla_bzl.diff",
+        "//openxla_patches:xplane.diff",
```
do we still need all of these patches after we moved to OpenXLA? I thought a lot of them are tf specified

ManfeiBai replied:
all the patches are applied to OpenXLA according to current TF-pin;

according to our branch `newbranchopxla`:
- `openxla_patches/optimized_function_graph.diff` is empty now, could delete this file
- `openxla_patches/xplane.diff` is empty now, could delete this file

other patch file would need confirmed before delete

ManfeiBai replied:
since https://github.com/pytorch/xla/pull/5219 has passed all tests, will apply this change later

JackCaoG commented on configuration.yaml:
```diff
@@ -113,6 +81,21 @@ variables:
         - Max inflight computations that the PJRT client can handle for TPU.
       type: int
       default_value: 32
+    RT_SHARD_LOCAL_ORDINAL:
+      description:
+        - Ordinal to be appended to the paths used by this thread of the xla
+          client.
+      type: int
+    RT_SHARD_ORDINAL:
+      description:
+        - Sets the shard ordinal for the XRT computation client.
+      type: int
+      default_value: -1
+    RT_WORKERS:
+      description:
+        - Addresses for the RT workers to be used by the RT client, for
+          example localservice:0;grpc://localhost:51011
+      type: string
```
I think auto-replace hit some bug here

ManfeiBai replied:
sry, didn't follow, what kind of bug?

JackCaoG replied:
you don't need `RT_WORKERS` and `RT_SHARD_ORDINAL`, they should be removed

JackCaoG replied:
Did you remove the logic to acquire these env var? I think they are just some XRT thing that can be removed.

ManfeiBai replied:
thanks for clarification, will delete these three envs `RT_SHARD_LOCAL_ORDINAL`, `RT_SHARD_ORDINAL` and `RT_WORKERS` after rebase with #5240

will-cromar replied:
It looks like these got added back in the merge. Please remove the `RT` config vars.

ManfeiBai replied:
thanks, updated

JackCaoG commented on docker/common.sh:
```diff
@@ -1,8 +1,7 @@
 function run_deployment_tests() {
-  export XRT_DEVICE_MAP="CPU:0;/job:localservice/replica:0/task:0/device:XLA_CPU:0"
-  export XRT_WORKERS="localservice:0;grpc://localhost:40934"
+  export PJRT_DEVICE=TPU
```
maybe `PJRT_DEVICE=TPU`? the above command is to configure XRT CPU

ManfeiBai replied:
Thanks for pointing, test in https://github.com/pytorch/xla/pull/5220

ManfeiBai replied:
#5220's change has been saved in https://github.com/pytorch/xla/pull/5222 too, since https://github.com/pytorch/xla/pull/5222 has passed all tests, will apply this change in #5220

ManfeiBai replied:
Thanks, updated

JackCaoG commented on openxla_patches/thread_local_random.diff:
```diff
@@ -0,0 +1,40 @@
+https://github.com/pytorch/xla/issues/4084
```
if I remember correctly this is a XRT related patch we can just remove...

ManfeiBai replied:
thanks for pointing, test in https://github.com/pytorch/xla/pull/5221

ManfeiBai replied:
#5221's change has been saved in https://github.com/pytorch/xla/pull/5222 too, since https://github.com/pytorch/xla/pull/5222 has passed all tests, will apply this change in #5221

ManfeiBai replied:
Thanks, updated

JackCaoG commented on openxla_patches/grpc_version.diff:
```diff
@@ -1,64 +1,8 @@
 # Tensorflow's patch to update the protobuf will regress the XRT runtime.
```
this is for XRT only

ManfeiBai replied:
thanks for pointing, test in https://github.com/pytorch/xla/pull/5222

ManfeiBai replied:
https://github.com/pytorch/xla/pull/5222 has passed all tests, will apply this change

ManfeiBai replied:
Thanks, updated

JackCaoG commented on test/pjrt/test_runtime.py:
```diff
@@ -55,24 +52,21 @@ def test_xla_device_error(self):
       # Print a warningif we had to select a default runtime
       if 'PJRT_DEVICE' not in os.environ and expect_using_pjrt:
         logs_context = self.assertLogs(level=logging.WARNING)
+        print("XRT has been deprecated, please set PJRT_DEVICE")
```
you should just error out here, PJRT will become the only runtime, not setting it is an error

ManfeiBai replied:
thanks for pointing, test in https://github.com/pytorch/xla/pull/5223

ManfeiBai replied:
try to raise an error here, but failed in https://github.com/pytorch/xla/pull/5223/commits/cd0f293c586e55e75cd3b573241d3ca9533cd2b0

since this is a test to test pjrt,  we have several situation here for PJRT test:
-   ('default', /*env_vars=*/{}, /*expect_using_pjrt=*/True), 
  - this set is testing whether we have PJRT as default
-   ('no_default', /*env_vars=*/{'PJRT_SELECT_DEFAULT_DEVICE': '0'}, /*expect_using_pjrt=*/False), 
  - this set is no PJRT_DEVICE set, and no device get
-  ('pjrt_cpu', /*env_vars=*/{'PJRT_DEVICE': 'CPU', 'PJRT_SELECT_DEFAULT_DEVICE': '0'}, /*expect_using_pjrt=*/True), 
  - PJRT_DEVICE=CPU test, will keep
-  ('xrt_tpu', /*env_vars=*/{'XRT_TPU_CONFIG': 'localservice;0;localhost:51011'}, /*expect_using_pjrt=*/False), 
  - this set has been deleted, no XRT code and no XRT TPU test
-  ('pjrt_tpu_precedence', /*env_vars=*/{'PJRT_DEVICE': 'TPU', 'XRT_TPU_CONFIG': 'localservice;0;localhost:51011',}, /*expect_using_pjrt=*/True), 
  - PJRT_DEVICE=TPU test, will keep
-  ('xrt_gpu', /*env_vars=*/{'GPU_NUM_DEVICES': '4'}, /*expect_using_pjrt=*/False), 
  - this set has been delelted, no XRT code and no XRT GPU test
-  ('pjrt_gpu', /*env_vars=*/{'PJRT_DEVICE': 'GPU','GPU_NUM_DEVICES': '4'}, /*expect_using_pjrt=*/True), 
  - PJRT_DEVICE=GPU test, will keep
-  ('xla_dist_worker', /*env_vars=*/{'XRT_LOCAL_WORKER': 'c_localservice:2'}, /*expect_using_pjrt=*/False))
  - this set will be delete, `xla_dist` is for XRT, and will delete XRT's xla_dist test


for situation which PJRT is not set and still want PJRT, we will set default PJRT_DEVICE to `CPU` or `TPU` based on whether or not found libtpu.so and TPU device: https://github.com/pytorch/xla/blob/b3f4c7bc1e9f812069770c83e38e5bd2dc20aef6/torch_xla/runtime.py#L36C1-L51C41

so we should not raise ERROR here for each situation

ManfeiBai replied:
Thanks, updated by deleting this print comments

JackCaoG commented on test/run_tests.sh:
```diff
@@ -108,42 +108,6 @@ function run_xla_backend_mp {
-  XLA_GET_TENSORS_OPBYOP=1 XLA_SYNC_TENSORS_OPBYOP=1 run_xrt "$@"
-}
-
-function run_async_scalar {
-  echo "Running in Async Scalar Upload mode: $@"
-  XLA_TRANSFER_SCALAR_ASYNC=1 run_xrt "$@"
-}
-
-function run_torchrun {
-  echo "Running tests spawned by torchrun"
-  if [ -x "$(command -v nvidia-smi)" ]; then
-    run_xrt "$@"
-  else
-    echo "the tests need atleast two XLA workers to validate"
-  fi
-}
```
why are these removed?

ManfeiBai replied:
`run_async_scalar` is only used in `run_xrt_tests`,
`run_torchrun` is only used in `run_xrt_tests` too :
 https://github.com/pytorch/xla/blob/b3f4c7bc1e9f812069770c83e38e5bd2dc20aef6/test/run_tests.sh#LL138C10-L138C23 

so deleted `run_async_scalar` and `run_torchrun` when delete `run_xrt_tests`

JackCaoG commented on test/test_xla_dist.py:
```diff
@@ -587,7 +587,7 @@ def test_unhealthy_pod_service_cluster(self):
             'state':
                 'READY',
             'health':
-                'UNHEALTHY_TENSORFLOW',
```
test_xla_dist can be removed, `xla_dist` only works for xrt

ManfeiBai replied:
thanks for pointing, test in https://github.com/pytorch/xla/pull/5224

ManfeiBai replied:
since #5224 has passed all tests, will apply this change

ManfeiBai replied:
Thanks, updated

JackCaoG commented on torch_xla/core/xla_env_vars.py:
```diff
@@ -1,17 +1,15 @@
-WORKERS = 'XRT_WORKERS'
-LOCAL_ORDINAL = 'XRT_SHARD_LOCAL_ORDINAL'
-ORDINAL = 'XRT_SHARD_ORDINAL'
+WORKERS = 'RT_WORKERS'
+LOCAL_ORDINAL = 'RT_SHARD_LOCAL_ORDINAL'
+ORDINAL = 'RT_SHARD_ORDINAL'
```
do we still need these? A lot of them are used by XRT only. can you find out who set these env var?

ManfeiBai replied:
these var are added 4 years ago:
https://github.com/pytorch/xla/blame/b3f4c7bc1e9f812069770c83e38e5bd2dc20aef6/torch_xla/core/xla_env_vars.py#LL7-L9C30

could create a ticket to track whether we still need these env var or not

ManfeiBai replied:
has rebased and these veriables has been deleted before rebase

JackCaoG commented on torch_xla/core/xla_model.py:
```diff
@@ -140,6 +140,7 @@ def get_ordinal(defval=0):
   if runtime.using_pjrt():
     return runtime.global_ordinal()
 
+  print("Please set PJRT_DEVICE")
```
I think you can just error out here. @will-cromar to confirm

ManfeiBai replied:
thanks for pointing, test in https://github.com/pytorch/xla/pull/5228

ManfeiBai replied:
Thanks, updated

JackCaoG commented on torch_xla/csrc/runtime/env_vars.cc:
```diff
@@ -7,16 +7,13 @@ namespace env {
 const char* const kEnvMeshService = "XRT_MESH_SERVICE_ADDRESS";
 const char* const kEnvWorldSize = "XRT_SHARD_WORLD_SIZE";
 const char* const kEnvMpDevice = "XRT_MULTI_PROCESSING_DEVICE";
-const char* const kEnvHostOrdinal = "XRT_HOST_ORDINAL";
-const char* const kEnvShardOrdinal = "XRT_SHARD_ORDINAL";
-const char* const kEnvStartService = "XRT_START_LOCAL_SERVER";
```
again I suspect we don't need these XRT_* env var, @will-cromar  to confirm

ManfeiBai replied:
thanks, will track in ticket too

ManfeiBai replied:
Thanks, updated

JackCaoG commented on torch_xla/csrc/runtime/metrics_reader.h:
```diff
@@ -14,7 +14,7 @@ namespace metrics_reader {
 // Creates a report with the current metrics statistics.
 std::string CreateMetricReport(
-    const std::map<std::string, torch_xla::runtime::Metric>& xrt_metrics);
+    const std::map<std::string, torch_xla::runtime::Metric>& rt_metrics);
```
you can just call it `metrics` instead of `rt_metrics`

ManfeiBai replied:
thanks for pointing, test in https://github.com/pytorch/xla/pull/5229

ManfeiBai replied:
Thanks, updated

JackCaoG commented on torch_xla/csrc/runtime/topology.proto:
```diff
@@ -0,0 +1,49 @@
+}
+
+// Describes the geometry of a TPU mesh.
+message TopologyProto {
```
why do we need this? We should be able to remove mesh service after XRT. PJRT does not use mesh service.

ManfeiBai replied:
PJRT has a test for mesh service: https://github.com/pytorch/xla/blob/master/test/pjrt/test_mesh_service.py

not sure whether this test is still needed or not

ManfeiBai replied:
mesh service would been deleted in https://github.com/pytorch/xla/pull/5240#pullrequestreview-1495708204, will delete `topology.proto` after rebase with merged #5240

ManfeiBai replied:
has rebased to Jun27, topology has been deleted before rebase

JackCaoG commented on torch_xla/distributed/xla_backend.py:
```diff
@@ -42,13 +41,6 @@ def __init__(self, prefix_store, rank, size, timeout):
-    # creating two TCP stores. We only want to call this
-    # when the user is using torchrun and not xmp.spawn()
-    # or some other flow.
-    if os.getenv('TORCHELASTIC_RUN_ID') != None:
```
Note, we should give AWS folks a heads up when we merge this pr.

ManfeiBai replied:
got it, have added in the list of full list of OpenXLA migration

JackCaoG commented on torch_xla/distributed/xla_multiprocessing.py:
```diff
@@ -55,7 +55,7 @@ def _create_gpu_devices(num_gpus):
 
 def _parse_workers_config(config):
-  # XRT_WORKERS='worker:0;ismz9:25822'
+  # PJRT_DEVICE=TPU
```
a lot of functions in this file is only used by XRT, we should do a clean up afterward.

JackCaoG replied:
@will-cromar to review change in this file

ManfeiBai replied:
sure, will do and saved in the list of full list of OpenXLA miration list

ManfeiBai replied:
has revert this change and @will-cromar will have follow-up PR for XRT contain this change

will-cromar commented:
Thanks @ManfeiBai. I'm very excited to see your progress here!

There are three separate changes that I would like to be submitted separately:

1. Delete XRT from the main branch
2. Swap TensorFlow and OpenXLA
3. Update pin to the latest XLA and libtpu

(2) and (3) can happen in any order after (1)

All three of these changes have significant risks. I expect each one to require a rollback for unforeseen issues, and I'd like to be able to roll back each step incrementally if needed.

I'll prepare a separate PR for (1) that deletes XRT since I have a good idea of where the dead code will be.

ManfeiBai commented:
> Thanks @ManfeiBai. I'm very excited to see your progress here!
> 
> There are three separate changes that I would like to be submitted separately:
> 
> 1. Delete XRT from the main branch
> 2. Swap TensorFlow and OpenXLA
> 3. Update pin to the latest XLA and libtpu
> 
> (2) and (3) can happen in any order after (1)
> 
> All three of these changes have significant risks. I expect each one to require a rollback for unforeseen issues, and I'd like to be able to roll back each step incrementally if needed.
> 
> I'll prepare a separate PR for (1) that deletes XRT since I have a good idea of where the dead code will be.

Thanks, @will-cromar, will rebase with your (1) PR after it merged

ManfeiBai commented:
fixed according to comments in this PR and https://github.com/pytorch/xla/pull/5222

will rebase with https://github.com/pytorch/xla/pull/5240/files later

ManfeiBai commented on bazel/openxla.bzl:
```diff
@@ -1,8 +1,8 @@
-"""Macros for working with tensorflow deps."""
```
delete this file after rebased with#5240

ManfeiBai replied:
rebased and keep it for a short time, would delete in the future

ManfeiBai replied:
testing in https://github.com/pytorch/xla/pull/5260

ManfeiBai commented:
check `rt_metrics` and `GetMetricRenderInfo()` in `[torch_xla/csrc/runtime/metrics_reader.cc](https://github.com/pytorch/xla/pull/5222/files/5b94b8b4ec33bab9eac08a90cb26af659ffcc3e4#diff-5c3a4f94c1a72a1ad2de9ad677e8f288ccaf1afe01108eddb771daa927b571f8)` according to @miladm's comment in https://github.com/pytorch/xla/pull/5222#discussion_r1239121138

ManfeiBai commented:
`CreateXrtMetricReport` is deleted and `GetMetricRenderInfo` is kept for now, leave this as open discuss for now, and get back to it after rebased with #5240

ManfeiBai commented:
rebased #5240 with and testing in https://github.com/pytorch/xla/pull/5242, once all tests passed in #5242, will update @5202 too

ManfeiBai commented on docs/bazel.md:
````diff
@@ -10,13 +10,6 @@ Tensorflow is a [bazel external dependency](https://bazel.build/external/overvie
 ```bzl
 http_archive(
     name = "org_tensorflow",
-    patch_args = [ "-l", "-p1"],
````
we have deleted these two patches in this PR, so keep these two patches in this md might confuse;

but this .md file is not updated at all, so we should update this document in a follow-up, and that would be clear too

ManfeiBai commented on torch_xla/csrc/runtime/sys_util.h:
```diff
@@ -17,7 +17,7 @@ std::string GetEnvOrdinalPath(const char* name, const std::string& defval,
 
 std::string GetEnvOrdinalPath(
     const char* name, const std::string& defval,
-    const char* ordinal_env = "XRT_SHARD_LOCAL_ORDINAL");
```
thanks, updated

ManfeiBai commented on torch_xla/csrc/runtime/metrics_analysis.cc:
```diff
@@ -90,76 +90,6 @@ class MetricTime : public Analyzer {
   long threshold_nsec_;
 };
 
-class XrtMetricFrequency : public Analyzer {
```
Thanks, updated and testing

ManfeiBai commented on test/pjrt/test_runtime.py:
```diff
@@ -8,25 +8,25 @@
 import torch_xla.core.xla_env_vars as xenv
 import torch_xla.core.xla_model as xm
-from torch_xla import runtime as xr
+from torch_xla import runtime as rt
```
thanks, updated

ManfeiBai replied:
thanks, updated

ManfeiBai commented on docker/common.sh:
```diff
@@ -1,8 +1,7 @@
 #!/bin/bash
```
thanks, updated

ManfeiBai commented on torch_xla/distributed/xla_dist.py:
```diff
@@ -0,0 +1,696 @@
+#!/usr/bin/env python
```
thanks, updated and testing in https://github.com/pytorch/xla/pull/5260

ManfeiBai commented on torch_xla/distributed/xrt_init.py:
```diff
@@ -0,0 +1,249 @@
+import os
```
thanks, updated and testing in https://github.com/pytorch/xla/pull/5260

will-cromar reviewed (approved):
LGTM

ddunl commented:
Nice work!

JackCaoG commented on torch_xla/csrc/runtime/openxla_exported_symbols.lds:
```diff
@@ -1,4 +1,3 @@
-*tensorflow*
 *absl*
```
I have a feeling that a lot of these won't be needed anymore, but we can clean it up later

ManfeiBai replied:
sure, thanks for pointing, will add in the list too, will have follow-up PR to update doc and test to delete these lds files too

ManfeiBai replied:
This file has been deleted in https://github.com/pytorch/xla/pull/5275

JackCaoG reviewed (approved):
Thanks @ManfeiBai ! This looks great! Let's try to merge this pr tmr!
